### ISSUE TITLE
[bug] Packages using msys2 as a build requirement use wrong compiler when compiling for Mingw64

### ISSUE THREAD

**Issue by syncu2** [conan-io/conan#11509]
I am using msys2 with the mingw-w64 toolchain as a build system and have noticed a problem with several conan packages using msys2 as a build requirement, e.g. libiconv and bison. 

Here is a minimum example using CMake:

CMakeLists.txt:
```
cmake_minimum_required(VERSION 3.9)
project(TestProject CXX)

list(APPEND CMAKE_MODULE_PATH ${CMAKE_BINARY_DIR})
list(APPEND CMAKE_PREFIX_PATH ${CMAKE_BINARY_DIR})

if(NOT EXISTS "${CMAKE_BINARY_DIR}/conan.cmake")
  message(STATUS "Downloading conan.cmake from https://github.com/conan-io/cmake-conan")
  file(DOWNLOAD "https://raw.githubusercontent.com/conan-io/cmake-conan/0.18.1/conan.cmake"
                "${CMAKE_BINARY_DIR}/conan.cmake"
                TLS_VERIFY ON)
endif()

include(${CMAKE_BINARY_DIR}/conan.cmake)

conan_cmake_configure(REQUIRES libiconv/1.16
                      GENERATORS cmake_find_package)

conan_cmake_autodetect(settings)

conan_cmake_install(PATH_OR_REFERENCE .
                    BUILD libiconv/1.16
                    REMOTE conancenter
                    SETTINGS ${settings})


add_executable(main main.cpp)
```

main.cpp:
```
int main(int argc, char** argv) {
    return 0;
}
```

I installed MSYS2 under G:\msys64\ and with pacman, installed the binutils for mingw64. Therefore, the compiler binaries are located in G:\msys64\mingw64\bin\.
To build, I execute the following steps from the MSYS2 Mingw64 (**not** MSYS) shell:

```
mkdir build
cd build
cmake -G "Ninja" -DCMAKE_BUILD_TYPE=Release ..
```

Conan then installs its own msys2 and after that, builds libiconv in it:

`libiconv/1.16: run_in_windows_bash: C:\.conan\a58c4a\1\bin\msys64\usr\bin\bash.exe --login -c ^"cd \^"/c/users/admin/.conan/data/libiconv/1.16/_/_/build/7e4fe5e99aec5cc8f78973ab8ce64613ed35677c/source_subfolder\^" ^&^& PATH=\^"/c/.conan/a58c4a/1/bin/msys64/usr/bin:$PATH\^" ^&^& make -j16 ^"`

The problem: The wrong gcc compiler is used, because the msys2 installed by Conan already ships with its own MSYS gcc toolchain which is cygwin-based. Due to it adding /c/.conan/a58c4a/1/bin/msys64/usr/bin to the PATH before calling make, this wrong compiler is now used when executing "gcc".
Wrong compiler: C:\.conan\a58c4a\1\bin\msys64\usr\bin\gcc.exe (MSYS/cygwin-based compiler)
Right compiler: G:\msys64\mingw64\bin\gcc.exe (MINGW64 compiler)

As a result, the library does not compile:
```
stat-time.h: In function �get_stat_atime_ns�:
stat-time.h:52:43: error: invalid use of undefined type �const struct _stati64�
   52 | #  define STAT_TIMESPEC(st, st_xtim) ((st)->st_xtim)

```
because the library is not meant to be built inside the MSYS/cygwin environment.

Expected behavior: Conan should recognize that it is being run inside a msys2 MINGW64 environment and set the compiler paths accordingly before calling its own msys2.

**Comment by memsharded** [comment]
Hi @syncu2 

There are too many moving pieces here, lets try to reduce the scope, lets try:

- Better not use recipes from ConanCenter, they can be too complex. Lets start with a library with ``conan new dep/0.1 --template=cmake_lib``, then an executable like ``conan new app/0.1 --template=cmake_exe`` and modifying it to add a ``requires = "dep/0.1"``, and use the code from ``dep/0.1``
- Lets drop the ``conan.cmake`` cmake-conan integration, as it adds an extra layer that would hide and limit control

Then, if packages are intended for the msys2 environment, it is very important to set the ``os.subsystem=msys2`` setting in the profile. Lets try to build the dependency first, then the app requiring it, lets see how it goes.

**Comment by syncu2** [comment]
Hello @memsharded 
thanks for helping me out. Not sure if you understood me right, though. I don't want to create a new recipe. I am building a private application which depends on libiconv/1.16 and simply want to build and install that dependency using Conan.

To be clear: By msys2, I mean both my installation of msys2 (from https://www.msys2.org/) which I use to execute cmake and the msys2 installation which Conan automatically installs to build libiconv. There seems to be a conflict between the two, as I stated above.

**Comment by memsharded** [comment]
Yes, I got that. But it is possible that ``libiconv/1.16`` has some issues. In that case, we would need to transfer the issue to ``conan-center-index`` repo, in case the recipe could be improved. Maybe it is not possible, ConanCenter recipes cannot target all the possible variants out there, and they are testing in Windows for Visual Studio compiler, and they might need to use ``msys2`` as a build env to run autotools tools, but the target compiler is still Visual Studio one.

So the first step is to validate if we can build very simple recipes in that environment, removing the specifics and complexity of ``libiconv/1.16``. If we manage to have something that works nicely, we have better chances that this could be used in ConanCenter. Otherwise, it is likely that this issue will not be prioritized in ConanCenter and be buried in the thousands of issues and pull requests there.

**Comment by syncu2** [comment]
Yes, your first paragraph makes perfect sense and matches my observations. Visual Studio toolchain works fine, because the compiler is called cl.exe and it uses some auxiliary script to invoke the compiler rather than calling it directly and therefore it does not conflict with the gcc.exe from msys2. 

In fact, I also tested the bison package which also uses msys2. Same issue there. 
I will test it with your method, tomorrow.

**Comment by syncu2** [comment]
With `conan new dep/0.1 --template=cmake_lib`, the build failed as well for the same reason as before.

However, I managed to solve the problem by installing MinGW separately (outside MSYS) and creating a conan profile with CC and CXX explicitly set to the compiler path:

```
[settings]
os=Windows
os_build=Windows
arch=x86_64
arch_build=x86_64
compiler=gcc
compiler.version=12.1
compiler.libcxx=libstdc++11
build_type=Release
[options]
[build_requires]
[env]
CC=D:/mingw64/bin/gcc.exe
CXX=D:/mingw64/bin/g++.exe
```

I was not aware that you are supposed to set those, as Cmake and Conan automatically determine the compiler information based on the current PATH and this works for all the packages which are not using the msys bash to build. I am fine with setting CC/CXX manually, but maybe, a warning message when they are not set and msys is invoked, would be very helpful, as this issue is very hard to track down.

**Comment by AbrilRBS** [comment]
Closing as Conan 1-only, which is no longer supported. The profile here uses `os_build`/`arch_build`, settings that don't exist in Conan 2's model (build-machine info now comes from the build profile / `settings_build` instead). If a similar msys2/Mingw64 compiler mixup happens with the two-profile setup on Conan 2, please open a new issue.
